iRODS 5: Restore eviction logic for expired dns/hostname caches, etc. (main)

korydraughn opened the pull request:
The important changes of this PR are:
- DNS/Hostname caches are now being cleared
  - This step fell through the cracks during the initial redesign of the startup/shutdown process
- Optional config properties in iRODS 4 are now required
- Environment variables defined in server_config.json are now being set
  - Child processes inherit the values

Running tests. Will open issues once the tests are passing.

trel commented on server/main_server/src/main.cpp:
```diff
@@ -280,11 +274,17 @@ auto main(int _argc, char* _argv[]) -> int
         log_server::info("{}: Initializing shared memory for main server process.", __func__);
 
         namespace hnc = irods::experimental::net::hostname_cache;
```
interested in pulling these out of the experimental namespace?

also still noticed 'log' is in experimental too...

korydraughn replied:
Yes, but I'm afraid we may run out of time to complete that effort since the blast radius is wide. I'm mostly speaking about the logger. The hostname and dns cache are pretty simple to handle.

I'll open issues for that effort so we can circle back to them after everything for iRODS 5 is confirmed to be working.

korydraughn replied:
Created #8399, #8400, #8403, and a few others for various libs.

alanking commented on lib/core/src/rcMisc.cpp:
```diff
@@ -4684,7 +4684,8 @@ int get_canonical_name(const char *_hostname, char* _buf, size_t _len)
 
     if (CLIENT_PT != ::ProcessType) {
-        const std::chrono::seconds age{irods::get_dns_cache_eviction_age()};
+        const auto cache_config = irods::get_advanced_setting<nlohmann::json>(irods::KW_CFG_DNS_CACHE);
+        const std::chrono::seconds age{cache_config.at(irods::KW_CFG_EVICTION_AGE_IN_SECONDS).get<int>()};
```
I realize string allocations can throw and we are leaning on the schema validation, but I think we should handle exceptions around this.

korydraughn replied:
Done.

alanking commented on lib/core/src/rcMisc.cpp:
```diff
@@ -4725,7 +4726,8 @@ int load_in_addr_from_hostname(const char* _hostname, struct in_addr* _out)
 
     if (CLIENT_PT != ::ProcessType) {
-        const std::chrono::seconds age{irods::get_dns_cache_eviction_age()};
+        const auto cache_config = irods::get_advanced_setting<nlohmann::json>(irods::KW_CFG_DNS_CACHE);
+        const std::chrono::seconds age{cache_config.at(irods::KW_CFG_EVICTION_AGE_IN_SECONDS).get<int>()};
```
I realize string allocations can throw and we are leaning on the schema validation, but I think we should handle exceptions around this.

korydraughn replied:
Agreed. The original code handled exceptions and this should do the same.

korydraughn replied:
Done.

alanking commented on lib/core/src/sockComm.cpp:
```diff
@@ -489,7 +489,8 @@ sockOpenForInConn( rsComm_t *rsComm, int *portNum, char **addr, int proto ) {
                 }
 
-                const std::chrono::seconds age{irods::get_hostname_cache_eviction_age()};
+                const auto cache_config = irods::get_advanced_setting<nlohmann::json>(irods::KW_CFG_HOSTNAME_CACHE);
+                const std::chrono::seconds age{cache_config.at(irods::KW_CFG_EVICTION_AGE_IN_SECONDS).get<int>()};
```
I realize string allocations can throw and we are leaning on the schema validation, but I think we should handle `nlohmann::json` exceptions and/or `std::exception` around this.

korydraughn replied:
Agreed. The original code handled exceptions and this should do the same.

korydraughn replied:
Done.

alanking commented on lib/core/include/irods/irods_server_properties.hpp:
```diff
@@ -282,43 +282,6 @@ namespace irods
-    /// \retval Configured-Value Otherwise.
-    ///
-    /// \since 4.2.9
-    auto get_dns_cache_shared_memory_size() noexcept -> int;
```
I'm guessing we don't need to deprecate these because these are pretty deep in the server despite being in `lib`?

korydraughn replied:
The plan is to mark them deprecated in 4.3.5. An issue for removing these in 5.0.0 would appear as well. I delayed creating the issues to make sure the work could be completed on time.

korydraughn replied:
Opened #8417 and #8418.

alanking reviewed (commented):
Makes sense overall.

korydraughn commented on server/main_server/src/main.cpp:
```diff
@@ -1461,6 +1464,38 @@ Environment Variables:
 
+    auto evict_expired_dns_cache_entries() -> void
+    {
+        const auto cache_config = irods::get_advanced_setting<nlohmann::json>(irods::KW_CFG_DNS_CACHE);
+        const auto sleep_time = cache_config.at(irods::KW_CFG_CACHE_CLEARER_SLEEP_TIME_IN_SECONDS).get<int>();
+
+        static auto start_time = std::chrono::steady_clock::now(); 
+        const auto now = std::chrono::steady_clock::now();
+
+        if (now - start_time < std::chrono::seconds{sleep_time}) {
+            return;
+        }
+
+        start_time = now;
+        irods::experimental::net::dns_cache::erase_expired_entries();
```
Consider exceptions.

korydraughn replied:
Done.

korydraughn commented on server/main_server/src/main.cpp:
```diff
@@ -1461,6 +1464,38 @@ Environment Variables:
+
+    auto evict_expired_hostname_cache_entries() -> void
+    {
+        const auto cache_config = irods::get_advanced_setting<nlohmann::json>(irods::KW_CFG_HOSTNAME_CACHE);
+        const auto sleep_time = cache_config.at(irods::KW_CFG_CACHE_CLEARER_SLEEP_TIME_IN_SECONDS).get<int>();
+
+        static auto start_time = std::chrono::steady_clock::now(); 
+        const auto now = std::chrono::steady_clock::now();
+
+        if (now - start_time < std::chrono::seconds{sleep_time}) {
+            return;
+        }
+
+        start_time = now;
+        irods::experimental::net::hostname_cache::erase_expired_entries();
```
Consider exceptions.

korydraughn replied:
Done.

korydraughn commented:
All unit tests passed.

All core tests passed except the following:
- `test_delay_queue.Test_Delay_Queue`
- `test_rule_engine_plugin_framework.Test_Plugin_Instance_CppDefault`

Investigating those failures now.

korydraughn commented:
The test failures appear to be timing related. I've seen the tests pass multiple times now.

I did discover that a cpp-default REP test was leaving an intermediate replica in storage, so I fixed that. I also discovered that many of the delay queue tests do not target a specific REP. I'll open an issue to look into that outside of this PR.

Going to address review comments now.

alanking commented:
> I did discover that a cpp-default REP test was leaving an intermediate replica in storage, so I fixed that.

We should probably create an issue to address that. It means the L1 descriptor is opened but the servicing agent never closes it. I see that `msiDataObjCreate` is calling `rsDataObjCreate` (which is just a wrapper around `rsDataObjOpen`), so it somehow is getting around the safeguard of the agent handling the close.

korydraughn commented:
That's true. Will see if I can reproduce that intermediate state.

korydraughn commented:
After some investigation, I'm certain the failures are due to timing.

The cpp-default REP test uses a `delay_assert` function (similar to `lib.delayAssert`) which only gives the passed callable 5 chances to succeed. The test passes when `ils` lists a replica in the intermediate state or stale state. The test fails whenever the replica isn't shown. With that said, I'm sure things are working (i.e. agents are cleaning up correctly) and the test just needs to be tighten up. This is further supported by the fact that the test session collections are removed without error. If an intermediate replica was present, we'd see error messages during test teardown.

I have not been able to reproduce the test failure I saw with the delay queue tests, but looking at the source for that test suite, it's easy to see that it can be tightened up as well. There are many cases where `irule` is invoked without a specific REP.

The test failures are not caused by the changes in this PR. I'm going to open issues for those test suites and focus on addressing the remaining review comments.

korydraughn commented:
Created #8414 and #8415 for the test failures.

korydraughn commented:
Unit tests still pass. No need to rerun the core test suite since the latest changes just add try-catch blocks and tweak one python test.

korydraughn commented:
This is ready for final review.

alanking reviewed (commented):
Makes sense to me... Awaiting test results.

I say go ahead and squash to your liking.

korydraughn commented:
> Makes sense to me... Awaiting test results.

See https://github.com/irods/irods/pull/8398#issuecomment-2822339268.

alanking commented:
Oh, I misread that. Yeah, go ahead and squash

korydraughn commented:
Squashed.

alanking reviewed (approved):
\# it

korydraughn commented:
Applied clang-format.

Squashed and pounded.

korydraughn commented:
Post merge update: Decided to rerun the core tests. All core tests passed.
